### DISCUSSION TITLE
Автоматическая подписка на альбомы

### DISCUSSION THREAD

**Discussion by amd64fox** [Chimildic/goofy#164] [Личный опыт]
```javascript
function subscribeAlbum() {

    const LIKES = 5;  // Кол-во лайков в альбоме для добавления в подписки

    // Не подписываться на эти альбомы
    const excludeAlbums = [
        //  { id: '' }, 
        //  { id: '' },  
    ];

    let loveTracks = Source.getSavedTracks();
    let follow_AlbumTracks = Source.getSavedAlbumTracks();

    let albumTracks = [];
    excludeAlbums.forEach((album) => Combiner.push(albumTracks, Source.getAlbumTracks(album)));
    Combiner.push(follow_AlbumTracks, albumTracks);
    Filter.removeTracks(loveTracks, follow_AlbumTracks);

    let unfollowAlbums = findUnfollowAlbums();
    // Library.saveAlbums(unfollowAlbums);  // Раскомментировать это, что бы начал подписываться

    function findUnfollowAlbums() {
        let unfollowAlbums = Object.values(calcRepeat())
            .filter((r) => r.count >= LIKES)
            .map((r) => r.item);
        if (unfollowAlbums.length == 0) {
            console.log('Альбомы не найдены');
        }
        if (unfollowAlbums.length >= 1) {
            console.log('Альбомов без подписки:', unfollowAlbums.length);

            // Показать имя альбомов 
            // console.log(unfollowAlbums.map((item) => item.name));

            // Показать URI альбомов
            // console.log(unfollowAlbums.map((item) => item.uri));
        }
        return unfollowAlbums;
    }

    function calcRepeat() {
        let loveAlbums = loveTracks.map((t) => t.album);
        return loveAlbums.reduce((repeat, album) => {
            repeat[album.id] = repeat[album.id] || { count: 0, item: album };
            repeat[album.id].count++;
            return repeat;
        }, {});
    }
}  
```

**Comment by efremych** [comment]
Я такой вот функцией пользуюсь. Добавляю по треку с альбома в плейлист и баню альбомы. С тем же успехом можно и альбомы передавать. 

Например
`Filter.removeAlbums(tempTracks, banAlbums, 'tracks');` - исключить все треки альбомов, из которых в массиве треки. 
`Filter.removeAlbums(artistAlbums, banAlbums);` - исключить альбомы. 

```js 
function removeAlbums(sourceArray, removedArray, type, invert = false) {
    let removedIds = removedArray.map((item) => item.album.id);
    let filteredTracks = [];
    
    if (!type) {
      filteredTracks = sourceArray.filter((item) => {
        return invert ^ !removedIds.includes(item.id);
      });
    };

    if (type == 'tracks') {
      filteredTracks = sourceArray.filter((item) => {
        return invert ^ !removedIds.includes(item.album.id);
      });
    };
    
    Combiner.replace(sourceArray, filteredTracks);
  } 
```

P.S. Все эти махинации с альбомами (дискавериальбом например) предполагают, что в библиотеке сохранены треки именно с альбома, а не из синглов. сборников и прочее. Не самая точная методика на мой взгляд... На поверхности, да, но не юзабильная. Тоже думал такое сделать, но пока не добрался, надо думать как сделать это действительно универсальным, иначе, что, треки у себя пересохранять и следить за библиотекой? :)

**Reply by amd64fox** [reply]
Благодарю за пример, буду тестить.

> Не самая точная методика на мой взгляд... На поверхности, да, но не юзабильная.

Если можно подробнее, о чем здесь идет речь ?

**Reply by efremych** [reply]
> Если можно подробнее, о чем здесь идет речь ?

В данном случае у меня должен быть сохранен в библиотеке хотя бы 1 трек с альбома, иначе откуда появится информация об этом альбоме. А ведь может быть что в любимых треках у меня 5 треков из синглов и все эти треки присутствуют в одном альбоме исполнителя, но счётчик любимых треков для данного альбома будет равен 0. 

В идеале вытаскивать все треки-альбомы исполнителя и по названию трека из любимых искать в каком альбоме этот трек есть. Но это довольно затратно по запросам :)

**Reply by amd64fox** [reply]
Аа, вы продолжаете мысль про синглы и сборники, тупанул, да да, есть такая небольшая проблема, но возможно и её можно исправить, видел в веб сервисе `skiley` фишку под названием `нормализация`

<details>
<summary><small>скрин с  её описанием</small></summary><p>

![Снимок экрана 2022-01-04 182522](https://user-images.githubusercontent.com/62529699/148082225-50fcbd45-e159-4429-ad5e-1e589b3dc1fa.jpg)
</details>

Тут конечно не совсем то что нужно, но суть похожа, если разработчик нашел способ обновить старый альбом на новый, то думаю можно также сделать с синглами и с сборниками, жаль, что он скрыл свой проект.

**Comment by amd64fox** [comment]
Блин, скрипт переделал на прошлой неделе, забыл скинуть, обновил первый пост, единственно я бы изменил источник любимых треков, я беру из кеша, не уверен, что всем это надо поэтому по умолчанию оставил `Source.getSavedTracks`
Для теста отключил подписку, что бы было видно при первом запуске скрипта на что будет подписываться, мало ли че, для боевых условий нужно раскомментировать `// Library.saveAlbums(unfollowAlbums);`
